sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsClientBuilder.java

```diff
@@ -0,0 +1,221 @@
+ * #buildAsyncClient() buildAsyncClient} respectively to construct an instance of the desired client.
+ */
+@ServiceClientBuilder(serviceClients = {DigitalTwinsClient.class, DigitalTwinsAsyncClient.class})
+public class DigitalTwinsClientBuilder {
+    private final List<HttpPipelinePolicy> additionalPolicies = new ArrayList<>();
+    // mandatory
+    private String endpoint;
+    private TokenCredential tokenCredential;
+    // optional/have default values
+    private DigitalTwinsServiceVersion serviceVersion;
+    private HttpPipeline httpPipeline;
+    private HttpClient httpClient;
+    private HttpLogOptions logOptions;
+    private RetryPolicy retryPolicy;
+
+    private static HttpPipeline buildPipeline(TokenCredential tokenCredential, String endpoint,
+                                              HttpLogOptions logOptions, HttpClient httpClient,
+                                              List<HttpPipelinePolicy> additionalPolicies, RetryPolicy retryPolicy) {
+        // Closest to API goes first, closest to wire goes last.
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+
+        // Adds a "x-ms-client-request-id" header to each request. This header is useful for tracing requests through Azure ecosystems
+        policies.add(new RequestIdPolicy());
+
+        // Only the RequestIdPolicy will take effect prior to the retry policy
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+        policies.add(retryPolicy);
+
+        // Adds a date header to each HTTP request for tracking purposes
+        policies.add(new AddDatePolicy());
+
+        // Add authentication policy so that each HTTP request has authorization header
+        HttpPipelinePolicy credentialPolicy = new BearerTokenAuthenticationPolicy(tokenCredential, GetAuthorizationScopes(endpoint));
+        policies.add(credentialPolicy);
+
+        policies.addAll(additionalPolicies);
+
+        // Custom policies, authentication policy, and add date policy all take place after the retry policy
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
+
+        policies.add(new HttpLoggingPolicy(logOptions));
+
+        return new HttpPipelineBuilder()
+            .policies(policies.toArray(new HttpPipelinePolicy[0]))
+            .httpClient(httpClient)
+            .build();
+    }
+
+    private static String[] GetAuthorizationScopes(String endpoint) {
+        // Uri representation for azure digital twin app Id "0b07f429-9f4b-4714-9392-cc5e8e80c8b0" in the public cloud.
+        String adtPublicCloudAppId = "https://digitaltwins.azure.net";
+        String defaultPermissionConsent = "/.default";
+
+        // If the endpoint is in Azure public cloud, the suffix will have "azure.net" or "ppe.net".
+        // Once ADT becomes available in other clouds, their corresponding scope has to be matched and set.
+        if (endpoint.indexOf("azure.net") > 0
+            || endpoint.indexOf("ppe.net") > 0) {
```

Comment:
Given this will be called a lot you may want to generate a `Pattern` for this.

```java
Pattern.compile("(ppe|azure)\\.net")
```
Could also add case-insensitive flag to it if that is allowed.